In [2]:
import numpy as np
import pandas as pd
from yelpapi import YelpAPI
from googleplaces import GooglePlaces, types, lang
import googlemaps 
import gmaps as maps

In [3]:
# Open Yelp API key file and read key
# The file is assumed to contain only one single line with the API key

def read_file_content(filename):
    """ Given a filename,
        returns the content of this file
    """
    try:
        with open(filename, 'r') as f:
            return f.read()
    except FileNotFoundError:
        print("'%s' file not found" % filename)

YELP_API_KEY = read_file_content("./API Keys/YELP_API_KEY.txt")
yelp_api = YelpAPI(YELP_API_KEY, timeout_s=3.0)

In [4]:
# Configure Google Places
PLACES_API_KEY = read_file_content("./API Keys/GOOGLE_PLACES_API_KEY.txt")
gmaps = googlemaps.Client(key=PLACES_API_KEY)

### Yelp Top 20 vs. Google Top 20 Restaurants in Stuttgart

Comparing the Top 20 results for Restaurants near Stuttgart from Google vs. Yelp, it seems that the Google ranking tends to prefer high-quality restaurants while the Yelp ranking tends to prefer restaurants visited frequently (e.g. 'Frittenwerk' vs. 'Hotel Royal', 'Biertargen im Schlossgarten' vs. 'Der Zauberlehrling')

In [5]:
# Lat/lng Stuttgart Hbf
lat_hbf = 48.783333
lng_hbf = 9.183333

In [6]:
# Top 20 restaurants in Stuttgart
response = yelp_api.search_query(term='restaurant', latitude=lat_hbf, longitude=lng_hbf, sort_by='best_match', limit=20)

In [7]:
def print_restaurant_names(response):
    restaurants = response['businesses']
    for rest in restaurants:
        print (rest['name'])
        
def get_restaurant_names(response):
    restaurant_list = []
    restaurants = response['businesses']
    for rest in restaurants:
        restaurant_list.append(rest['name'])
    return restaurant_list

In [8]:
print_restaurant_names(response)

Frittenwerk
Biergarten im Schlossgarten
Poffers Café Stuttgart
Gaststätte Schlesinger
Die Zirbelstube
Kleinschmeckerei
Takeshii's
Aspendos
Ramen 8
Gravity
Zum Becher
Italiani
Food Court
L'Osteria
Mandu
Yorma's
Rhythms Bar & Kitchen
Cube
Brauhaus Schönbuch
Injeera


In [11]:
# Compare Top 20 Google results with top 20 Yelp results -> there is no common element
top_20_yelp = get_restaurant_names(response)

# Google Top 20, which we recieved earlier (but saved only the IDs) 
top_20_google = ['Hotel Royal', 'Der Zauberlehrling', 'Alte Kanzlei', 'Oggi', 'Enchilada Stuttgart', 
                 '5 Bar Gourmetrestaurant', 'Valle', 'Sky Beach Stuttgart', 'BLOCK HOUSE Eberhardstraße', 
                 'Arche Weinstube', 'VAPIANO Suttgart Bolzstrasse', 'Weinstube Fröhlich', 
                 'Paulaner am alten Postplatz', 'Prince of India', 'Hotel-Restaurant Köhler', 'Ambiente Africa', 
                 'MAREDO Steakhaus Stuttgart', 'BLOCK HOUSE Arnulf-Klett-Platz', 'CUBE Restaurant', 'Amici']

list(set(top_20_yelp) & set(top_20_google))

[]

### Download Yelp IDs based on Multiple Google Places Nearby Searches

Earlier we downloaded a list of ~1300 Google Place IDs for the most popular restaurants in Stuttgart area. The list was obtained by making multiple Google Places Nearby Searches around varying search centers. This list of IDs represents a good selection of restaurants (in terms of relevancy and quality of the restaurants), and the starting point for our Restaurant database.

Since we are not allowed to save any Google Places data other than the ID, we need to gather the restaurant information from the Yelp API, matching the Google Places ID to the corresponding Yelp IDs.

To do so, we first look up name and location at Google Places for each Google ID, then try to find the corresponding Yelp business by making a Yelp search_query based on name and location.

In [13]:
# For each restaurant id in the Google list, lookup name and geolocation at Google,
# then do a matching query at Yelp

google_restaurant_ids = np.array(
    pd.read_csv('./data/google_restaurant_ids.csv')['id'])

In [19]:
# For each Google ID, look up lat, lng, and name at Google Places
# Try to find matching Yelp Business based on name and geolocation.

yelp_ids_based_on_google = []
for google_id in google_restaurant_ids:
    print("Find Yelp ID for Google ID: {}".format(google_id))
    try:
        # Lookup name + geolocations at Google by ID 
        google_restaurant = gmaps.place(place_id=google_id)
        lat = google_restaurant['result']['geometry']['location']['lat']
        lng = google_restaurant['result']['geometry']['location']['lng']
        name = google_restaurant['result']['name']
        print("...restaurant {} at {}, {}".format(name, lat, lng))
        # Now try to find it at Yelp based on name and geolocation
        yelp_restaurant = yelp_api.search_query(term=name, latitude=lat, longitude=lng, radius=500, limit=2)
        #print(yelp_restaurant)
        yelp_id = yelp_restaurant['businesses'][0]['id']
        #print(yelp_restaurant['businesses'][0]['coordinates'])
        yelp_latitude = yelp_restaurant['businesses'][0]['coordinates']['latitude']
        yelp_longitude = yelp_restaurant['businesses'][0]['coordinates']['longitude']
        yelp_ids_based_on_google.append(yelp_id)
        print("...Yelp ID: {}".format(yelp_id))
    except Exception:
        try:
            print("No match found for Google name: {}".format(name))
        except Exeption:
            print("No match found for Google id: {}.format(google_id)")
        continue

Find Yelp ID for Google ID: ChIJdQ5X5XzPmUcRv7wL1Ond7Go
No match found for Google name: Daily Burger
Find Yelp ID for Google ID: ChIJCyT7XoDPmUcRVlr-BH_bNw8
...restaurant Restaurant "La Perla" im Schießsport-Zentrum at 48.8333999, 9.249574
No match found for Google name: Restaurant "La Perla" im Schießsport-Zentrum
Find Yelp ID for Google ID: ChIJleGdJZDJmUcRqozTDacuv14
...restaurant Andys Crepes at 48.8319935, 9.258655899999999
No match found for Google name: Andys Crepes
Find Yelp ID for Google ID: ChIJjzY04XzFmUcRkoCkdy647Co
...restaurant Restaurant TSV Steinhaldenfeld at 48.82968539999999, 9.2403472
...Yelp ID: -k0Ksk_FSzQAiKU6tdjayg
Find Yelp ID for Google ID: ChIJbc5V5XzPmUcRHvMxfGR2Afw
...restaurant Oettingers Restaurant at 48.8340312, 9.2634533
...Yelp ID: -lWPadzpW-DFB0C6UHvX6w
Find Yelp ID for Google ID: ChIJqxZJe5vFmUcRl6JkPcV7kGk
...restaurant Die Früchtle vom Schmidener Feld at 48.81933780000001, 9.254840100000001
...Yelp ID: 7Qvx3vZzuTNtB7kbCbQqZA
Find Yelp ID for Google 

No match found for Google name: Pizza Trullo‘s
Find Yelp ID for Google ID: ChIJM5ZhX4zPmUcRwLzsyq0PDmI
...restaurant Arschlöchle at 48.8398046, 9.232103100000002
No match found for Google name: Arschlöchle
Find Yelp ID for Google ID: ChIJL-TuR_bPmUcRDA0wgxOZTjI
...restaurant Treffpunkt am See at 48.83564000000001, 9.21872
...Yelp ID: noXKdbLQqpZPAiesHqPzdw
Find Yelp ID for Google ID: ChIJq0yCCtfEmUcRJeLQt23fGOY
...restaurant Taverne Timbales at 48.8413185, 9.2357325
No match found for Google name: Taverne Timbales
Find Yelp ID for Google ID: ChIJB52jMvPPmUcRAE__uR8-qqU
...restaurant Grillstation at 48.84043639999999, 9.2290019
No match found for Google name: Grillstation
Find Yelp ID for Google ID: ChIJYZpz2I3PmUcRmxwTc6Y-AuQ
...restaurant Stuttgart-Hofen at 48.84128, 9.23577
No match found for Google name: Stuttgart-Hofen
Find Yelp ID for Google ID: ChIJYZpz2I3PmUcRGNHLiqVQjlg
...restaurant Vereinsgaststätte SKG Stuttgart at 48.84128, 9.23577
No match found for Google name: Vereinsgas

No match found for Google name: Am Eck
Find Yelp ID for Google ID: ChIJMSQ6N_zPmUcRqGfXcDTU46M
...restaurant Bäckerei Böhm GmbH at 48.8416691, 9.2083277
No match found for Google name: Bäckerei Böhm GmbH
Find Yelp ID for Google ID: ChIJ40x44gPQmUcR1YX2QW2foN0
...restaurant Gasthaus Krone at 48.8410636, 9.1981596
No match found for Google name: Gasthaus Krone
Find Yelp ID for Google ID: ChIJt4_uNUrFmUcRv6_SqW5xbLo
...restaurant Maxs Gastronomie GmbH at 48.81775, 9.2008
No match found for Google name: Maxs Gastronomie GmbH
Find Yelp ID for Google ID: ChIJlZCHLMnbmUcRtH_jgDKRYG0
...restaurant Istanbul Kebab at 48.8317786, 9.1867054
No match found for Google name: Istanbul Kebab
Find Yelp ID for Google ID: ChIJBfOl5KPbmUcRMckqqqaAfeM
...restaurant مطعم الجنة at 48.8304524, 9.1864378
No match found for Google name: مطعم الجنة
Find Yelp ID for Google ID: ChIJVVVljaXamUcR_CMrFL_4ero
...restaurant Tanzcafe at 48.8304524, 9.1864378
No match found for Google name: Tanzcafe
Find Yelp ID for Googl

No match found for Google name: Aslan Kebap-Haus
Find Yelp ID for Google ID: ChIJadP-RYTamUcRvcG5DMKbDJM
...restaurant Pasquale Pepe Gaststätte at 48.8322519, 9.157652299999999
No match found for Google name: Pasquale Pepe Gaststätte
Find Yelp ID for Google ID: ChIJqWPHA8DamUcRhCIhw0xyS30
...restaurant Göktaş Restaurant at 48.81748, 9.170589999999999
No match found for Google name: Göktaş Restaurant
Find Yelp ID for Google ID: ChIJF5x_csDamUcRUf1WjH0ecew
...restaurant Köfteci Tuncay Stuttgart at 48.8171451, 9.1717659
No match found for Google name: Köfteci Tuncay Stuttgart
Find Yelp ID for Google ID: ChIJs2olT5namUcRx2SmpVXHGcw
...restaurant Fäßle at 48.8328068, 9.1697089
No match found for Google name: Fäßle
Find Yelp ID for Google ID: ChIJ47Ev1p3amUcR6nwe3JHMOzw
...restaurant Weinstube zum Tröpfle at 48.83598509999999, 9.169606
...Yelp ID: mZ_q2M-TvBJe2DLKvkoRmQ
Find Yelp ID for Google ID: ChIJh2q_Y_PbmUcRgtCdcrV1u20
...restaurant Mersin Tantuni at 48.82982599999999, 9.17067229999999

No match found for Google name: Vereinsheim Dietel
Find Yelp ID for Google ID: ChIJbQLj4kfbmUcRQS-gV5qk4JQ
...restaurant Indian Grill Station at 48.83704710000001, 9.149160400000001
No match found for Google name: Indian Grill Station
Find Yelp ID for Google ID: ChIJi63cq3_amUcRRnPCNxTeP94
...restaurant BIG BABA’s Kebabhouse at 48.83769139999999, 9.1477416
No match found for Google name: BIG BABA’s Kebabhouse
Find Yelp ID for Google ID: ChIJVRqHNPzbmUcRngld6xkc3DA
...restaurant Miroslav at 48.8301201, 9.136284300000002
No match found for Google name: Miroslav
Find Yelp ID for Google ID: ChIJb_Y6H3ramUcRlIAwzg-LxSk
...restaurant Rustical Vereinsheim der Gartenfreunde Grenzäcker at 48.83537339999999, 9.138789
No match found for Google name: Rustical Vereinsheim der Gartenfreunde Grenzäcker
Find Yelp ID for Google ID: ChIJBWy913HamUcRUkvrjkJYyJA
...restaurant Milano Lieferservice at 48.831228, 9.131924900000001
No match found for Google name: Milano Lieferservice
Find Yelp ID for Google I

No match found for Google name: Kolaha Gastro GmbH
Find Yelp ID for Google ID: ChIJ3TSaLenFmUcRRx5rdUzLEDE
...restaurant Bistro 33 - carpe diem at 48.8157116, 9.259349900000002
No match found for Google name: Bistro 33 - carpe diem
Find Yelp ID for Google ID: ChIJEw69RpTFmUcRcgU4FYMm73o
...restaurant Pronto Pizza da Stefano at 48.8097234, 9.254286899999999
No match found for Google name: Pronto Pizza da Stefano
Find Yelp ID for Google ID: ChIJRaNEaZfFmUcRJnJxOX8Q0Lk
...restaurant JET Tankstelle at 48.81480000000001, 9.2602
No match found for Google name: JET Tankstelle
Find Yelp ID for Google ID: ChIJn3ZqnqDFmUcRS3TPtF9Ems4
...restaurant Britan Crêpes & Falafel Spezialitäten at 48.8084, 9.24444
No match found for Google name: Britan Crêpes & Falafel Spezialitäten
Find Yelp ID for Google ID: ChIJcVJuIYXFmUcRhzgBWdTVjfE
...restaurant Cannstatter Combüse at 48.8083124, 9.244042
No match found for Google name: Cannstatter Combüse
Find Yelp ID for Google ID: ChIJ87-xxJbFmUcR2Vf6I2WAh4U
...r

No match found for Google name: Theaterschiff Stuttgart GmbH & Co. KG
Find Yelp ID for Google ID: ChIJq6rW0zvFmUcRO_kK9g6o5d0
...restaurant Pizza Service Puglia at 48.80924460000001, 9.209587200000001
No match found for Google name: Pizza Service Puglia
Find Yelp ID for Google ID: ChIJCWH9lDvFmUcR8sOWa3SKIhY
...restaurant Pizza Express Deluxe Stuttgart at 48.8082113, 9.2106095
No match found for Google name: Pizza Express Deluxe Stuttgart
Find Yelp ID for Google ID: ChIJg_RjAT3FmUcR7FTgRHRorwE
...restaurant Taverna Ouzeri Kolokotronis at 48.8063164, 9.214734100000001
...Yelp ID: LM-_1Ye4Ufn9zxEuZkIvEw
Find Yelp ID for Google ID: ChIJuQLqbj3FmUcRYWp4l1OlYp8
...restaurant The Little Greek Taverna at 48.805664, 9.216522
...Yelp ID: U6-pS4idHuRXjr3wpxUmIQ
Find Yelp ID for Google ID: ChIJQRu64jzFmUcR5QFrTOzrRsY
...restaurant Die Palette at 48.8061278, 9.214599800000002
...Yelp ID: XmC8Omoy8uG2krHmgFSBUg
Find Yelp ID for Google ID: ChIJpxm3hDvFmUcRSd7udR-GGb4
...restaurant Thai Viet - asian 

No match found for Google name: Bistro WerkstattKaffee
Find Yelp ID for Google ID: ChIJDwfq6NLamUcRQQBbfdGhhq0
...restaurant cafe fossil am Dinomuseum / Naturkundemusesum am Löwenstein at 48.8051644, 9.190527799999998
No match found for Google name: cafe fossil am Dinomuseum / Naturkundemusesum am Löwenstein
Find Yelp ID for Google ID: ChIJAROsl8bamUcRxYB7Ib81qvI
...restaurant Workstation at 48.8132682, 9.1759688
No match found for Google name: Workstation
Find Yelp ID for Google ID: ChIJ9_0GvEfdmUcR6KJ6UmxrGvo
...restaurant Schmücker Gastro & Catering GmbH at 48.81134, 9.175790000000001
No match found for Google name: Schmücker Gastro & Catering GmbH
Find Yelp ID for Google ID: ChIJpSjQ2dDEmUcRLDGRhyat7kA
...restaurant Lúc Lắc at 48.80857, 9.183442999999999
No match found for Google name: Lúc Lắc
Find Yelp ID for Google ID: ChIJSZS8Es_amUcRLMetA3xOxcI
...restaurant Pavillon am Pragsattel at 48.81003689999999, 9.1822255
No match found for Google name: Pavillon am Pragsattel
Find Yelp I

No match found for Google name: Istanbul kebab
Find Yelp ID for Google ID: ChIJ-3RaqenamUcRduhDROYRev4
...restaurant Kebaphaus at 48.8111105, 9.165396199999998
No match found for Google name: Kebaphaus
Find Yelp ID for Google ID: ChIJ811Y-uTamUcRBUt-o_l4vvE
...restaurant Körle und Adam at 48.805825, 9.156069400000002
...Yelp ID: -UfNsMlVYFAZZypRnxCQiA
Find Yelp ID for Google ID: ChIJa3Sv4u7amUcRIh2MecpZ3CU
...restaurant Panther Pizza Feuerbach at 48.80914689999999, 9.1570988
No match found for Google name: Panther Pizza Feuerbach
Find Yelp ID for Google ID: ChIJa3GTz-_amUcRNqjUTbLKLcQ
...restaurant Retsinadiko, die griechische Taverne at 48.8107838, 9.156018299999998
No match found for Google name: Retsinadiko, die griechische Taverne
Find Yelp ID for Google ID: ChIJ4WTlre_amUcRPpDUBNTyCRg
...restaurant Pizzeria Da Luigi at 48.80891999999999, 9.15934
...Yelp ID: SadLEDO0zJOGb0IxuhYjAg
Find Yelp ID for Google ID: ChIJDfhluO_amUcR74OYxBe_r3g
...restaurant Kirschbaum Feuerbach at 48.81083

No match found for Google name: Evangelisches Waldheim Lindental
Find Yelp ID for Google ID: ChIJvzE3bzjamUcR2vxhJ8Z1ESY
...restaurant IKOSUSHI at 48.8138864, 9.1163639
...Yelp ID: al6ZLQHhtPFQQrJqGfPAnw
Find Yelp ID for Google ID: ChIJ4W9aaDjamUcRJqL8BcdGrmY
...restaurant MAIA Asiatische Küche -Sushi at 48.8138834, 9.116008100000002
No match found for Google name: MAIA Asiatische Küche -Sushi
Find Yelp ID for Google ID: ChIJC1VISTjamUcR0zEIVp-qB-E
...restaurant Mevlana kebap at 48.81371619999999, 9.1146245
No match found for Google name: Mevlana kebap
Find Yelp ID for Google ID: ChIJ2YL1VMfoC0ER_QojXLslRZU
...restaurant Baran Kebab at 48.813733, 9.114443
...Yelp ID: CVS4vhkO9oUH0oQ6iJucaw
Find Yelp ID for Google ID: ChIJs8rWq0XamUcRFaR8E6RomCY
...restaurant Paros Restaurant at 48.81349909999999, 9.12674
...Yelp ID: KoaPhdiB_PDS1NHeDxn0aw
Find Yelp ID for Google ID: ChIJiQLrhmzbmUcR8AYp9eubw1s
...restaurant Schwabenstüble at 48.814182, 9.122054100000001
No match found for Google name: 

No match found for Google name: Meraki-Greek Taverna
Find Yelp ID for Google ID: ChIJAYXTvRzFmUcRE9T8Ilt4dwE
...restaurant Kulturinsel-Biergarten at 48.7973323, 9.2262678
No match found for Google name: Kulturinsel-Biergarten
Find Yelp ID for Google ID: ChIJuYF8DBjFmUcRvlsOOW6o_yM
...restaurant Nic-Nac at 48.80274, 9.21706
No match found for Google name: Nic-Nac
Find Yelp ID for Google ID: ChIJKfou0JLFmUcRgcm1eewbG-Q
...restaurant Ngon Deli - Vietnamese Home Cooking at 48.8015444, 9.222419400000001
No match found for Google name: Ngon Deli - Vietnamese Home Cooking
Find Yelp ID for Google ID: ChIJd0s2O_7EmUcRDyfxmIh8Zb0
...restaurant PSV at 48.7918211, 9.2345652
...Yelp ID: OkohZaV6gYCylMSeFxYYKA
Find Yelp ID for Google ID: ChIJw1CW0oTFmUcRRuPUe-fCedE
...restaurant MaM-MaM Hot Dog at 48.8016267, 9.2214176
No match found for Google name: MaM-MaM Hot Dog
Find Yelp ID for Google ID: ChIJkb0kqRnFmUcRV50rDQiT5ZI
...restaurant Grillhähnchen at 48.8017323, 9.2214001
No match found for Google 

No match found for Google name: Huong Viet - Vietnamesische Spezialitäten
Find Yelp ID for Google ID: ChIJLxGZpMDbmUcRdkm31DSL5FI
...restaurant Atelier kaifischerdotorg at 48.7994204, 9.185453899999999
No match found for Google name: Atelier kaifischerdotorg
Find Yelp ID for Google ID: ChIJt0XoBdTamUcRf6OfsaOqMQ4
...restaurant IBIS KITCHEN at 48.8011474, 9.184425599999999
No match found for Google name: IBIS KITCHEN
Find Yelp ID for Google ID: ChIJzRdgjhnFmUcRxwXjhgIFJ2M
...restaurant Subway at 48.7929129, 9.1904741
...Yelp ID: 5q8n_btjQu4nJOBvi_SVgw
Find Yelp ID for Google ID: ChIJPS9ssNfEmUcRioZ8FKKaYR0
...restaurant Kiosk im Schloßgarten at 48.7935891, 9.198400099999999
No match found for Google name: Kiosk im Schloßgarten
Find Yelp ID for Google ID: ChIJHa6V89fEmUcRNVBV2XHuyEU
...restaurant Begleitbüro SOUP e.V. at 48.794171, 9.200568500000001
No match found for Google name: Begleitbüro SOUP e.V.
Find Yelp ID for Google ID: ChIJza3DGkDFmUcRknzI6bOBAlY
...restaurant Fossil - Café, E

No match found for Google name: enjoy and eat
Find Yelp ID for Google ID: ChIJF7F3DUXEmUcRE-ahPEAPdBQ
...restaurant Glorious Pizza Hot & Spicy at 48.7805458, 9.250859699999998
No match found for Google name: Glorious Pizza Hot & Spicy
Find Yelp ID for Google ID: ChIJzwhY80XEmUcRwE4vHBBqYn0
...restaurant Sapori d'Italia at 48.78055, 9.25292
...Yelp ID: rMk0fOt65ixXxnL0eMLIEA
Find Yelp ID for Google ID: ChIJOZrqjEXEmUcR1EMHSqdgi9A
...restaurant Imbiss Meteora at 48.78048, 9.25274
No match found for Google name: Imbiss Meteora
Find Yelp ID for Google ID: ChIJv1J2fUXEmUcRxdpSy4IZwRc
...restaurant SPEISEREI at 48.7801902, 9.2512071
...Yelp ID: n4-b1WE8l7_5RITMXTcncw
Find Yelp ID for Google ID: ChIJ6-fC5UXEmUcRMAEEoSlkscM
...restaurant Bar Pizzeria da Matteo at 48.779886, 9.2535104
No match found for Google name: Bar Pizzeria da Matteo
Find Yelp ID for Google ID: ChIJczZyxEnEmUcRUePIJ53rjOk
...restaurant Stoighofbesen Kitzele at 48.7836818, 9.2605553
No match found for Google name: Stoighofb

No match found for Google name: Cancun
Find Yelp ID for Google ID: ChIJv8K_2vDEmUcRdQW59_AUtXA
...restaurant Restaurant Istanbul at 48.7808098, 9.228217599999999
No match found for Google name: Restaurant Istanbul
Find Yelp ID for Google ID: ChIJqaVBdFrFmUcR3g1s1x2PRJE
...restaurant AHI Burger & Kebap Grillhaus at 48.77891539999999, 9.232933599999999
No match found for Google name: AHI Burger & Kebap Grillhaus
Find Yelp ID for Google ID: ChIJ12KfO_HEmUcRd3JhxtQ34B4
...restaurant Ristorante La Villa at 48.77959330000001, 9.227332200000001
...Yelp ID: YEjtmVEvuS7dafksiFxsKA
Find Yelp ID for Google ID: ChIJy6nkXcrFmUcRPfnAK6tq2FI
...restaurant ESG Kutschers Glück at 48.779763, 9.224471399999999
No match found for Google name: ESG Kutschers Glück
Find Yelp ID for Google ID: ChIJNxo_q4XFmUcRJv0xnP0b3CU
...restaurant Hasen Stüble at 48.77974339999999, 9.224393899999997
No match found for Google name: Hasen Stüble
Find Yelp ID for Google ID: ChIJ__8_Rl_EmUcRXEynWZonCZ0
...restaurant Maridin R

No match found for Google name: FS Development GmbH
Find Yelp ID for Google ID: ChIJqa9KFtLEmUcR-5vi9KlemzQ
...restaurant Lek’s Thailändisches Spezialitätenrestaurant & Lounge at 48.786997, 9.191364099999998
No match found for Google name: Lek’s Thailändisches Spezialitätenrestaurant & Lounge
Find Yelp ID for Google ID: ChIJP2FZCdLEmUcRnyVrfCTxwds
...restaurant san&soup at 48.7861093, 9.1907208
No match found for Google name: san&soup
Find Yelp ID for Google ID: ChIJ6TOC2s3EmUcRZPHSiAbt3EY
...restaurant Poffers Café Stuttgart at 48.7839221, 9.190614000000002
...Yelp ID: c9e3voWpBlsANfWeZct1ZA
Find Yelp ID for Google ID: ChIJ05RykMDEmUcRjBN4hFTYqIc
...restaurant Jakova Premium Fast Food at 48.7753511, 9.203679200000002
No match found for Google name: Jakova Premium Fast Food
Find Yelp ID for Google ID: ChIJ0TK6MMzEmUcRSbAIFVXrbis
...restaurant PrinZ - Kebap Pizza Schawarma Falafel at 48.7828514, 9.189822699999997
No match found for Google name: PrinZ - Kebap Pizza Schawarma Falafel
Find

No match found for Google name: Bar Ristorante Fellini
Find Yelp ID for Google ID: ChIJ872vt0vbmUcRGEMcSRjiyf0
...restaurant Indian Palace at 48.7737031, 9.1767568
...Yelp ID: a4bWT9eFj9kUydQbg1xdkA
Find Yelp ID for Google ID: ChIJjTWMo0vbmUcRnEe4sZqVemU
...restaurant Cortijo at 48.7738013, 9.178107899999999
...Yelp ID: 8hJh2bS6yxJ7ld0KyhsRHA
Find Yelp ID for Google ID: ChIJWfDcTz_bmUcRVjDPwDDAmL8
...restaurant Wirtshaus zum Hotzenplotz at 48.77856, 9.16307
...Yelp ID: 56832hA2kNtvtRzV7cjBXA
Find Yelp ID for Google ID: ChIJma5uwknbmUcRczaV_7Pn14I
...restaurant Udo Snack at 48.7761642, 9.174099900000003
...Yelp ID: MN-wdtlFj6jPApv3ojZSvg
Find Yelp ID for Google ID: ChIJCTp-TTzbmUcRkkkK3qHB_iw
...restaurant Divan at 48.7845927, 9.1612029
...Yelp ID: tAayweHA0dCq3J0W3y6psg
Find Yelp ID for Google ID: ChIJN8YaQErbmUcRUmrCtbtK2ao
...restaurant Monroe's at 48.77559280000001, 9.177346199999999
...Yelp ID: 0MNos77iTyXNl5kzJHMEoA
Find Yelp ID for Google ID: ChIJW6msukvbmUcRfHMEypZs17E
...restau

...Yelp ID: yypxgWicbQIFw28UJnkkOA
Find Yelp ID for Google ID: ChIJ_3DuEezbmUcRUBumWe7kK-M
...restaurant Pizzeria at 48.7850324, 9.143768300000001
No match found for Google name: Pizzeria
Find Yelp ID for Google ID: ChIJHR0BtXXbmUcR4jrGOCKyqUE
...restaurant Bistro Benzin at 48.7821344, 9.1573915
No match found for Google name: Bistro Benzin
Find Yelp ID for Google ID: ChIJuyju-xXbmUcRNbc6uiKKS9A
...restaurant Da Livio e Alessia at 48.78305599999999, 9.1583504
No match found for Google name: Da Livio e Alessia
Find Yelp ID for Google ID: ChIJe_wVED3bmUcRoYdBO33-l1U
...restaurant Tre Amici at 48.7807061, 9.156806800000002
No match found for Google name: Tre Amici
Find Yelp ID for Google ID: ChIJ-ZWXiRXbmUcRPRf7CYFyw_s
...restaurant Hölderlin Platz at 48.78171810000001, 9.1582495
...Yelp ID: 9NXAnyJbamWe-_6ypOMO1w
Find Yelp ID for Google ID: ChIJD6BUC-_bmUcRiZutMIae_yY
...restaurant Kaek - Vegane Pâtisserie at 48.77967, 9.15719
No match found for Google name: Kaek - Vegane Pâtisserie
Find

No match found for Google name: Live - kebap & pizza
Find Yelp ID for Google ID: ChIJwxwvMGDEmUcR8eQIksGNmS8
...restaurant Friedrichsruh at 48.7702778, 9.2327778
...Yelp ID: KDIJf7YBdH434f0SXP6aRA
Find Yelp ID for Google ID: ChIJ-eQ7Cl3EmUcRi24mZaEdq4s
...restaurant Metzgerei Häfele (Stuttgart-Wangen) at 48.7731, 9.242510000000001
No match found for Google name: Metzgerei Häfele (Stuttgart-Wangen)
Find Yelp ID for Google ID: ChIJl620eWbEmUcRWKiq_B2fDQs
...restaurant Waldheim Wangen Restaurant Schembergers at 48.7667556, 9.241125000000002
No match found for Google name: Waldheim Wangen Restaurant Schembergers
Find Yelp ID for Google ID: ChIJr-4e5j3FmUcRk-iRuFjAJCQ
...restaurant Wangener Eckle Griechische Taverne at 48.77141229999999, 9.245858199999997
No match found for Google name: Wangener Eckle Griechische Taverne
Find Yelp ID for Google ID: ChIJcY3aPRPFmUcRM22FG4f6rMk
...restaurant Pizzabob at 48.76925980000001, 9.248285599999999
No match found for Google name: Pizzabob
Find Yelp ID

No match found for Google name: East Side Kebab
Find Yelp ID for Google ID: ChIJ3crqucPEmUcRZHpDZ2tfhL4
...restaurant El Koop at 48.7811264, 9.2055395
No match found for Google name: El Koop
Find Yelp ID for Google ID: ChIJe2GkbP3FmUcRznSXnk1pRqo
...restaurant Pizza Factory Stuttgart at 48.780367, 9.1998941
No match found for Google name: Pizza Factory Stuttgart
Find Yelp ID for Google ID: ChIJj-_SnH3FmUcRyvgEEbXQFVs
...restaurant Eat Happy at 48.7831365, 9.2075521
No match found for Google name: Eat Happy
Find Yelp ID for Google ID: ChIJ42hY18jEmUcRXUtvtD8xXJI
...restaurant Café Arg at 48.78000089999999, 9.1936801
No match found for Google name: Café Arg
Find Yelp ID for Google ID: ChIJFS9bGcnEmUcRSmIbvZQ2RXc
...restaurant Restaurant Lässig at 48.77912000000001, 9.19247
No match found for Google name: Restaurant Lässig
Find Yelp ID for Google ID: ChIJE0SAI0nbmUcRZenL1F_y3Sw
...restaurant Hotel Royal at 48.7732925, 9.172032399999999
...Yelp ID: abt4OGzqkisFz2533Iqskw
Find Yelp ID for G

No match found for Google name: Pizza Pasta Taxi - italienischer Lieferservice in Stuttgart-West mit Pizza, Pasta, veganen Gerichten und Salaten
Find Yelp ID for Google ID: ChIJsU4RiELbmUcRJuKwDp6Ifc8
...restaurant Jose y Josefina at 48.7708796, 9.155809300000001
...Yelp ID: zwSU147QTs9cGH9vmpSCyQ
Find Yelp ID for Google ID: ChIJyaR7FELbmUcReY7qtEYUb_Y
...restaurant Lumen at 48.77240489999999, 9.1566307
...Yelp ID: ZC6DED1nK9T6SfswWxCptw
Find Yelp ID for Google ID: ChIJN0RaAULbmUcRfFt9YXAkhBc
...restaurant Restaurant Achillion at 48.77139889999999, 9.157684899999998
...Yelp ID: Km1Ry4ZnNLeQdF8uC-dHAQ
Find Yelp ID for Google ID: ChIJU0HwVELbmUcRpB3R_P44BQ8
...restaurant Cesa Cigköfte at 48.7709519, 9.1579262
No match found for Google name: Cesa Cigköfte
Find Yelp ID for Google ID: ChIJFe0sdUPbmUcRM2whwvwvE8g
...restaurant QQ-Sushi Lounge at 48.770641, 9.159075999999997
...Yelp ID: 6CN4S6922egl2ebnV6BZeQ
Find Yelp ID for Google ID: ChIJu2zd9nrbmUcRcTdtNVLc5ws
...restaurant Vereinsheim Zi

No match found for Google name: Trattoria da noi toni&gigi
Find Yelp ID for Google ID: ChIJeYKLB3TEmUcRl5PyCK5WpME
...restaurant Gasthaus Zur Krone at 48.7580717, 9.2544199
No match found for Google name: Gasthaus Zur Krone
Find Yelp ID for Google ID: ChIJK5XwZNbFmUcRW-4E7aFsmQI
...restaurant Bistro Endstation at 48.75952330000001, 9.2542898
No match found for Google name: Bistro Endstation
Find Yelp ID for Google ID: ChIJLRLGTQ3EmUcRpYlYnxRfF68
...restaurant Bistro Hedelfinger Platz at 48.7596831, 9.254151600000002
No match found for Google name: Bistro Hedelfinger Platz
Find Yelp ID for Google ID: ChIJyYBC4QHnC0ERQH0k-0NtqRU
...restaurant Ulrich Bleibler at 48.7593674, 9.2546571
No match found for Google name: Ulrich Bleibler
Find Yelp ID for Google ID: ChIJgbEFGcnFmUcRh-Q435bidi0
...restaurant Gourmet Palast Stuttgart at 48.76090680000001, 9.2555459
No match found for Google name: Gourmet Palast Stuttgart
Find Yelp ID for Google ID: ChIJy1J-b3bEmUcRfsGfybhSLeU
...restaurant Hot-Chil

No match found for Google name: Kochschule Kochpunkt
Find Yelp ID for Google ID: ChIJszGNWFbDmUcR5psyGak-uKs
...restaurant Anna´s Waldau at 48.7535736, 9.183514299999999
No match found for Google name: Anna´s Waldau
Find Yelp ID for Google ID: ChIJ_UZN21bDmUcRqE8TtPcRGFo
...restaurant Ristorante-Pizzeria Masaniello at 48.7521744, 9.183306699999997
No match found for Google name: Ristorante-Pizzeria Masaniello
Find Yelp ID for Google ID: ChIJe68etsHDmUcRE60Y-HDuINM
...restaurant waldauerle at 48.75311910000001, 9.1818459
No match found for Google name: waldauerle
Find Yelp ID for Google ID: ChIJ___rClTbmUcR71QWcMik4s4
...restaurant Konkrua at 48.76188480000001, 9.1727412
No match found for Google name: Konkrua
Find Yelp ID for Google ID: ChIJdXTHMJnbmUcREmrUnRD9p7w
...restaurant TinTin at 48.76607490000001, 9.1784856
...Yelp ID: Wds_jUzsm6ydXugyRNZzXg
Find Yelp ID for Google ID: ChIJYzpqk1LbmUcRrDNq9fiAHnY
...restaurant Épicerie Fine at 48.7656216, 9.1767611
...Yelp ID: p1FPUxkQfSFqJGIk

No match found for Google name: Pizzeria La Mamma
Find Yelp ID for Google ID: ChIJl-Q3q17bmUcR_Rg63yRz5pM
...restaurant colombia candela at 48.7628707, 9.158700399999999
No match found for Google name: colombia candela
Find Yelp ID for Google ID: ChIJ8x4WJ1vbmUcRMl9m_rMfaw0
...restaurant Tschechen & Söhne at 48.7679701, 9.1653188
...Yelp ID: iwEvPMXK4l4zA2apLxOpdA
Find Yelp ID for Google ID: ChIJyYKgMFrbmUcRGDQ0YQOI3do
...restaurant New Asia Imbiss at 48.7636583, 9.1629556
...Yelp ID: 8fPpDaXnoSyotNeem4_pqA
Find Yelp ID for Google ID: ChIJsxnylV_bmUcR_oU50W90EMY
...restaurant Weinstube Heeb at 48.75971, 9.153229999999999
...Yelp ID: dkfr-WVOWODGK8lskFtaMg
Find Yelp ID for Google ID: ChIJAVSrma3cmUcRF5ipU_7B4Tk
...restaurant Café Russini at 48.7481535, 9.1694742
...Yelp ID: xF6AKTg4ZBIITwubGLgvfg
Find Yelp ID for Google ID: ChIJzWXFbKfcmUcRtcS0CS29M3A
...restaurant Vereinsgaststätte Gartenfreunde Stuttgart-Heslach at 48.75680000000001, 9.158920000000002
No match found for Google name: V

...Yelp ID: gsF8xaxbGX9AD-0FoZnu-g
Find Yelp ID for Google ID: ChIJ2UnbDjvDmUcR2WKRCevJ0S4
...restaurant Starbing Bubble Tea & Street Food 缘滋缘味 at 48.74140839999999, 9.269123000000002
No match found for Google name: Starbing Bubble Tea & Street Food 缘滋缘味
Find Yelp ID for Google ID: ChIJ3fDtb7rDmUcRw-Q86OEuxBY
...restaurant AMADEUS Eiscafé at 48.74140839999999, 9.269123000000002
No match found for Google name: AMADEUS Eiscafé
Find Yelp ID for Google ID: ChIJcxew6PzDmUcRT-zM2RFgns8
...restaurant Hähnchengrill at 48.74140839999999, 9.269123000000002
No match found for Google name: Hähnchengrill
Find Yelp ID for Google ID: ChIJcW-P8fzDmUcRqRRbaNoX1TA
...restaurant Asia Binh Minh at 48.7410405, 9.2692363
No match found for Google name: Asia Binh Minh
Find Yelp ID for Google ID: ChIJv0ztg3rDmUcRWG3hNjhR5UQ
...restaurant Ruiter Kebaphaus at 48.7299387, 9.2529536
No match found for Google name: Ruiter Kebaphaus
Find Yelp ID for Google ID: ChIJvbrKBzDnC0ERX3nBlzFBflM
...restaurant Kupferkessel 

...Yelp ID: 03RdFNkAXISiQ6sLndNRVg
Find Yelp ID for Google ID: ChIJ46pkTlrcmUcRfp0YhSSOu0M
...restaurant Harmonie Pizza Service at 48.7338454, 9.1084569
...Yelp ID: 81FR9Ncu_i9E8LBR3jh19g
Find Yelp ID for Google ID: ChIJq8i2M1zcmUcRlVKA4UDicDY
...restaurant Debitel QG Stuttgart at 48.7388055, 9.111433499999999
No match found for Google name: Debitel QG Stuttgart
Find Yelp ID for Google ID: ChIJj-edQFzcmUcRlTHHLGToBPQ
...restaurant TimeOut at 48.7377668, 9.110222199999999
No match found for Google name: TimeOut
Find Yelp ID for Google ID: ChIJq6rqG0LcmUcRg4w5SzgppWA
...restaurant SocialMania at 48.7415131, 9.1010301
No match found for Google name: SocialMania
Find Yelp ID for Google ID: ChIJidN06lfcmUcRDSHbSuqEFYQ
...restaurant Dong Kinh at 48.7299191, 9.1146959
...Yelp ID: cbqamMEVG5rPw4iSo6gGEQ
Find Yelp ID for Google ID: ChIJb8HIUVfcmUcR2u4hf1sZzz8
...restaurant Ya Ya Sushi at 48.7309915, 9.110281900000002
...Yelp ID: PuZNdoqk9XqGPjMeqqr_1g
Find Yelp ID for Google ID: ChIJX8BsAkLcmUc

No match found for Google name: Botticelli Food and More
Find Yelp ID for Google ID: ChIJeUjbwFDcmUcRZef-NDO0vdg
...restaurant asiahung at 48.73032800000001, 9.108488900000001
No match found for Google name: asiahung
Find Yelp ID for Google ID: ChIJlR_3ceDdmUcRa6Zuwh55UWg
...restaurant Lavanta Cafe&Bäckerei at 48.73007690000001, 9.103864500000002
No match found for Google name: Lavanta Cafe&Bäckerei
Find Yelp ID for Google ID: ChIJJZ8DgrPdmUcRv97o3BWmqmM
...restaurant Kebab Hütte Vaihingen Stuttgart Döner Pide Pizza Seele Drehspieß (Scheibenfleisch am Dienstag, Donnerstag, Samstag, Sonntag), (Hänchendöner am Montag, Mittwoch, Freitag), (Iskenderkebab am Donnerstag, Samstag, Sonntag) at 48.7300252, 9.1053864
No match found for Google name: Kebab Hütte Vaihingen Stuttgart Döner Pide Pizza Seele Drehspieß (Scheibenfleisch am Dienstag, Donnerstag, Samstag, Sonntag), (Hänchendöner am Montag, Mittwoch, Freitag), (Iskenderkebab am Donnerstag, Samstag, Sonntag)
Find Yelp ID for Google ID: ChIJ

No match found for Google name: Deutsch Cantina
Find Yelp ID for Google ID: ChIJSYImdg7DmUcR_K5_TUMnOAM
...restaurant Hunt Brothers Pizza at 48.72382650000001, 9.1761602
No match found for Google name: Hunt Brothers Pizza
Find Yelp ID for Google ID: ChIJPwaPe9PcmUcRLTe67mDKn2g
...restaurant SI-SUITES at 48.7239646, 9.161900500000002
...Yelp ID: czvBbsA02qiCyF5JlZPzfA
Find Yelp ID for Google ID: ChIJdX92ryzDmUcRT7PTos_X6GM
...restaurant Hotel Dormero at 48.7224317, 9.163261799999999
...Yelp ID: cnpyL8mCSEmjAC4QvRrApA
Find Yelp ID for Google ID: ChIJPwaPe9PcmUcRnjyOLuYe8mM
...restaurant Braustube Schlossturm at 48.7239092, 9.163105600000002
...Yelp ID: 6tR-uUoy8CScXuSRMw2igQ
Find Yelp ID for Google ID: ChIJPwaPe9PcmUcRqBqmQQYtFGk
...restaurant Red Grill Stuttgart at 48.72234520000001, 9.163226800000002
No match found for Google name: Red Grill Stuttgart
Find Yelp ID for Google ID: ChIJAfXM79PcmUcRScH0rV2UFHM
...restaurant Imbiss Inn at 48.72316389999999, 9.159522199999998
...Yelp ID: IEi

...Yelp ID: tdsVbIHvwPWFO764eWArvw
Find Yelp ID for Google ID: ChIJMxBCYifdmUcRjaIC5kvSIHc
...restaurant feinkost squisito at 48.71844, 9.146440000000002
No match found for Google name: feinkost squisito
Find Yelp ID for Google ID: ChIJv61J2dzcmUcRlxUdrJgyU-A
...restaurant Möhringer Pizza und Kebabhaus at 48.72787999999999, 9.144480000000001
No match found for Google name: Möhringer Pizza und Kebabhaus
Find Yelp ID for Google ID: ChIJZ7SxB9TdmUcRt4RGtDrdUUc
...restaurant Il Vulcano Ristorante Pizzeria at 48.72762729999999, 9.144417599999999
No match found for Google name: Il Vulcano Ristorante Pizzeria
Find Yelp ID for Google ID: ChIJhQtFi9zcmUcRwsLHnXp8S1M
...restaurant Momentos at 48.726394, 9.1453749
...Yelp ID: cJNw0OmXqYHumZsFFKtZFw
Find Yelp ID for Google ID: ChIJc7ND_DfdmUcR35dusO7jnUY
...restaurant Gaumenschmaus Haenchen at 48.7272853, 9.1413418
No match found for Google name: Gaumenschmaus Haenchen
Find Yelp ID for Google ID: ChIJqZzRkufcmUcR-_0a_XpPEV8
...restaurant Mediterra

No match found for Google name: Ivica Rezo Gaststättenbetriebe e.K.
Find Yelp ID for Google ID: ChIJVVUVqFfcmUcRORZYWoSeYyg
...restaurant Gülnaz Döner & Pizza at 48.7293539, 9.112712499999997
No match found for Google name: Gülnaz Döner & Pizza
Find Yelp ID for Google ID: ChIJf_1x5AvdmUcRD8Jh7REi_RM
...restaurant Eat Happy at 48.72962030000001, 9.1118479
...Yelp ID: 9A-y0TWBKqqP5vsn_CwI3g
Find Yelp ID for Google ID: ChIJB3Ey7FXcmUcRvQ2act8k2xk
...restaurant Aqua, Restaurant Bar und Lounge at 48.7251666, 9.1134439
No match found for Google name: Aqua, Restaurant Bar und Lounge
Find Yelp ID for Google ID: ChIJ0SO7gFbcmUcR91-ecxAdfhc
...restaurant FONTANA RESTAURANT at 48.7263747, 9.111349299999999
...Yelp ID: sj18RDxKq6n_JMYIJ8T4zg
Find Yelp ID for Google ID: ChIJib05d3vdmUcRyllsVjiP2fU
...restaurant Der Nudelstrauss ,,Lé Bouquet dé Pates" at 48.7302606, 9.111054700000002
No match found for Google name: Der Nudelstrauss ,,Lé Bouquet dé Pates"
Find Yelp ID for Google ID: ChIJK50s8iTdmUcR9

In [18]:
# Write to CSV file and Excel file
df = pd.DataFrame(yelp_ids_based_on_google, columns=['yelp_id'])
df.to_csv('yelp_ids_based_on_google_search.csv')
df.to_excel('yelp_ids_based_on_google_search.xlsx')

### Download Yelp IDs based on Multiple Yelp searches 

Only ~420 of the restaurants found via Google Places could be matched to a Yelp ID based on name and geolocation search. This is probably the case when the names at Google Places vs. Yelp differ too much (even though the Yelp API does not seem to require an exact match), or when the geolocations differ too much. However, we saw the results changing across multiple executions (cf. https://github.com/Yelp/yelp-fusion/issues/197), and the number of matches ranged between 408 and 430.

So we need to supplement these matches with additional results, retrieved from Yelp directly. 

Even though the Yelp API would support to request up to 1000 results around one location (instead 3x20 ), we prefer a grid-based search around multiple search centers (similar as we did with Google Places earlier), which is supposed to provide a higher coverage.

In [22]:
# Define inner and outer quadrangle within we want to search as well as the appropriate step width.
# For the inner quardrangle, we define a smaller step width. 

step = -0.029 # ~2km
smaller_step = step / 2

# Stuttgart area
start_lat = 48.85
end_lat = 48.66
start_long= 9.38
end_long = 8.95

# Stuttgart city
start_lat_city = 48.83
end_lat_city = 48.72
start_long_city = 9.25
end_long_city = 9.10


In [23]:
# Show search centers as markers on Google Maps

maps.configure(PLACES_API_KEY)
fig = maps.figure()

# Draw red markers for the wider area of Stuttgart
marker_locations=[]
for i in np.arange(start_lat, end_lat, step):
    for j in np.arange(start_long, end_long, step):
        # draw markers only if i and j is not within the city area
        if ((i > start_lat_city) | (i < end_lat_city) | (j > start_long_city) | (j < end_long_city)):
            marker_locations.append((i, j))
            
print("Number of red markers: {}".format(len(marker_locations)))   
markers_red = maps.symbol_layer(marker_locations, fill_color='blue', stroke_color='blue')
fig.add_layer(markers_red)

# Draw blue markers for the city area of Stuttgart
marker_locations=[]
for i in np.arange(start_lat_city, end_lat_city, smaller_step):
    for j in np.arange(start_long_city, end_long_city, smaller_step):
        marker_locations.append((i, j))

print("Number of blue markers: {}".format(len(marker_locations)))   
markers_blue = maps.symbol_layer(marker_locations, fill_color='purple', stroke_color='purple')
fig.add_layer(markers_blue)

fig


Number of red markers: 85
Number of blue markers: 88


Figure(layout=FigureLayout(height='420px'))

In [251]:
yelp_ids = []
yelp_names = []
yelp_lats = []
yelp_lngs = []
for marker in marker_locations:
    yelp_result = yelp_api.search_query(
        term='restaurant', latitude=marker[0], longitude=marker[1], radius=1500, limit=50)
    for i in range(len(yelp_result['businesses'])):
        yelp_id = yelp_result['businesses'][i]['id']
        name = yelp_result['businesses'][i]['name']
        lat = yelp_result['businesses'][i]['coordinates']['latitude']
        lng = yelp_result['businesses'][i]['coordinates']['longitude']
        yelp_ids.append(yelp_id)
        yelp_names.append(name)
        yelp_lats.append(lat)
        yelp_lngs.append(lng)
yelp_results.append(yelp_result)


In [254]:
assert(len(yelp_names) == len(yelp_ids) & len(yelp_ids) == len(yelp_lats) & len(yelp_lats) == len(yelp_lngs))

In [265]:
# Build DataFrame
d = {'yelp_id':yelp_ids, 'name':yelp_names, 'latitude': yelp_lats, 'longitude':yelp_lngs}
yelp_restaurants = pd.DataFrame(d)
yelp_restaurants.head()

yelp_id                                         name  \
0  -k0Ksk_FSzQAiKU6tdjayg               Restaurant Tsv Steinhaldenfeld   
1  5djuZfIkqf1ykQGbpWESrg                                       Stadio   
2  R93PaCrS4qai9u1uiiHPMw  Restaurant "La Perla" im Schießsportzentrum   
3  3e_HLr7Qu6Azg7Hp6EK3aw                                Rems La Pizza   
4  kscS_CV9KX-j1zdfvBu7mQ                                Il Carrettino   

    latitude  longitude  
0  48.829740   9.240350  
1  48.831900   9.252310  
2  48.833358   9.249487  
3  48.840622   9.267763  
4  48.833700   9.264480

In [268]:
yelp_restaurants['yelp_id'].nunique()

823

In [269]:
# De-dupliacte
yelp_restaurants = yelp_restaurants.drop_duplicates()

#### Yelp IDs found via Matching Google results vs. Yelp IDs found via Yelp directly

`yelp_ids_based_on_google` (IDs based on Google search): 
 * Based on the list of Google IDs, obtained by multiple Google Places Nearby Searches with varying search centers, returning 3x20 results each
 * Based on the Google ID, lookup the Google name and try to find a match at Yelp
 * For 1333 Google restaurants, only 430 matches could be identified, probalby because of name or geolocation mismatch between Yelp and Google (e.g. "Eat Drink Man Woman" vs. "Gasthaus Eat Drink Man Woman")

`yelp_ids_based_on_yelp` (IDs based on Yelp search): 
* Based on multiple Yelp Search Queries with 50 results each, using the same search grid as earlier for the Google search
* 823 results after de-duplication

`yelp_ids_based_on_google` - `yelp_ids_based_on_yelp`: 
* 97 restaurants that are in `yelp_ids_based_on_google`, but not in `yelp_ids_based_on_yelp` .
* Restaurants that were originally found via Google grid search, then mapped to the Yelp equivalents, but are not found via Yelp grid search

`yelp_ids_based_on_yelp` - `yelp_ids_based_on_google`: 
* 498 restaurants that are in `yelp_ids_based_on_yelp`, but not in `yelp_ids_based_on_google` 
* Restaurants that are found via Yelp grid search, but not included in the Yelp equivalents for our previous Google search
* Mostly due to 'name mismatch' between Google name and Yelp name 

==> We continue with a merged set (`yelp_ids_based_on_google` + `yelp_ids_based_on_yelp`) to avoid missing any relevant restaurant, wether it comes from Google or from Yelp. This gives us **823 + 97 = 920 restaurants in the area of Stuttgart**.

In [290]:
yelp_set = set(yelp_ids_based_on_yelp)
google_set = set(yelp_ids_based_on_google)
print(len(yelp_set), len(google_set))

823 422


In [336]:
# IDs that are included in Yelp-based grid search, but not in Google-based grid search
# This includes also the name mismatch cases
yelp_minus_google = yelp_set.difference(google_set)
len(yelp_minus_google)

498

In [335]:
# IDs that are included in Google-based search, but not in Yelp based search
google_minus_yelp = google_set.difference(yelp_set)
len(google_minus_yelp)

97

In [327]:
print("IDs contained in yelp-based but not in google-based search (mostly due to name mismatch)")
for id in yelp_minus_google:
    print(get_name(id))

IDs contained in yelp-based but not in google-based search
The Gardener’s Nosh
Ristorante Le Rose
Bäckerhaus Veit Café
Ristorante Pizzeria La Taverna da Enzo
Weinwirtschaft | Weingut Franz Keller
Zum Adler Gasthaus Weinstube
ESV Rot-Weiss Vereinsgaststätte
Restaurant Schmidener Eintracht
Bäckerhaus Veit Café
Mikail's Lunchbox
Da Mamma Maria
1/1
Yakiniku
Due Stanze e Cucina
Rhythms Bar & Kitchen
Goldener Adler
List
Ristorante Russo Wine Food and Eventlocation
Zum Becher
Gasthof Riedsee
Gasthaus zum Kreuz
Dam Dam Hot Pizza
SEN Asia Soul Kitchen
Cannstatter Tor
Gaststätte Post
Fischlounge
Geiler
Sattlerei
Il Terrazzino
Weinstube Kochenbas
Gaststätte Trompetle
Gravity
Triple B
Cafe & Bar Nion
Taverna Sultan Saray
ALATURKA - Das Stuttgarter Original
Kwan Kao
Noir Cuisine&Bar
[m]eatery
Campioni
Christo's Stube
Funfo's
Bahtti Super Pizza Service
Weinhalle 1896
Cavallino Spaghettaro
Diogenes Taverne
Flori & Palma
Vereinsgaststätte Himmerreich
Los Locos Latinos
Da Micci
Restaurant LE
le tonneau

In [332]:
print("IDs contained in google-based but not in yelp-based search")
for id in google_minus_yelp:
    print(get_name(id))

IDs contained in google-based but not in yelp-based search
Tokyo Sushi
Sultan Saray
Bocca Buona
Hotel Schmidener Eintracht
Cafe Le Theatre
Hotel Hirsch
Ristorante Positano
PAULANER am alten Postplatz
Gaststätte Trollinger
Arcotel Camino
Burger King
Schnauferl Grill
Burger House
Pizzeria Da Cesare
Le Cassoulet
Sultan Saray
Ebony
Aral
La Bruschetta
Amici
Art Sushi
Oggi
Icebreaker Bar im Hotel Pullman Fontana
Indian Palace
Hotalo
Pizza Fiesta
Ristorante La Piazza
Mauritius
Da Noi Due
Aral
JET
Logo - Rotisserie
La Commedia
Enchilada
Aral Tankstelle
Da Franco
Ceylonas
Spielbank Stuttgart
Mezzogiorno
SI-SUITES
Alte Kanzlei
Hotel Neuwirtshaus
Sausalitos
Bella Pizza Service
Eberhard Braun - Gourmetcook - Kochveranstaltungen
Claus
Akzent Hotel Möhringer Hof
5
The Dubliner
Indian Shop
1893 - Das VfB Clubrestaurant
DORMERO Hotel Stuttgart
Im Eimer
Laurita Pizza Service
Früchtle vom Schmidener Feld
Die Palette
Thai Corner
China-Imbiss-Panda
Weber
Pullman Stuttgart Fontana
Cafe Monroe's
Sakura
Lagu

In [361]:
def get_name(yelp_id):
    yelp_restaurant = yelp_api.business_query(id=yelp_id)
    return yelp_restaurant['name']
    

In [350]:
# Write IDs from Yelp-based search to file
df = pd.DataFrame(yelp_ids_based_on_yelp, columns=['yelp_id'])
df.to_csv('yelp_ids_based_on_yelp_search.csv', index=False)
df.to_excel('yelp_ids_based_on_yelp_search.xlsx', index=False)

In [351]:
# Save merged IDs from Google and Yelp-based search 
df = pd.DataFrame(list(merged_ids), columns=['yelp_id'])
df.to_csv('yelp_ids.csv', index=False)
df.to_excel('yelp_ids.xlsx', index=False)

### Enrich merged IDs with detail information

We now have a list of ~820 Yelp IDs, retrieved either by matching the Google Places to Yelp Businesses, or by doing Yelp searches directly. For these IDs, we need to gather the required information (mostly from Yelp). We perform the following steps: 
- For each Yelp ID, make a Yelp business request and enrich the given ID with detail information (such as name, categories, geolocations).
- For each Yelp ID, find the corresponding Google ID from Google Places. This is done with a Google Places find_place search based on Yelp name and location. For every Yelp ID, the corresponding Google ID can be found.

In [465]:
# Read in restaurant IDs.
df_yelp_ids = pd.read_csv('yelp_ids.csv')

In [466]:
# For each Yelp ID, get detail information from Yelp

id_list = []
name_list = []
is_closed_list = []
categories_list = []
price_list = []
rating_list = []
review_count_list = []
photos_list = []
address_list = []
lat_list = []
lng_list = []
for yelp_id in df_yelp_ids['yelp_id'].values:
    restaurant = yelp_api.business_query(id=yelp_id)
    name = restaurant['name']
    is_closed = restaurant['is_closed']
    categories = restaurant['categories']
    categories_unstacked = []
    for c in categories:
        categories_unstacked.append(c['title'])
    if 'price' in restaurant:
        price_level = restaurant['price']
    else:
        price_level = ''
    rating = restaurant['rating']
    review_count = restaurant['review_count']
    if 'photos' in restaurant:
        photo_urls = restaurant['photos']
    else: photo_urls = ''
    address = restaurant['location']['display_address']
    lat = restaurant['coordinates']['latitude']
    lng = restaurant['coordinates']['longitude']
    # Append everything to the lists
    id_list.append(yelp_id)
    name_list.append(name)
    address_list.append(address)
    is_closed_list.append(is_closed)
    categories_list.append(categories_unstacked)
    price_list.append(price_level)
    rating_list.append(rating)
    review_count_list.append(review_count)
    photos_list.append(photo_urls)
    lat_list.append(lat)
    lng_list.append(lng)
    
assert(len(id_list) == len(name_list) & len(name_list) == len(address_list) & len(address_list) == len(is_closed_list)
       & len(is_closed_list) == len(categories_list) & len(categories_list) == len(price_list) & 
       len(price_list) == len(rating_list) & len(rating_list) == len(review_count_list) & 
       len(review_count_list) == len(photos_list) & len(photos_list) == len(lat_list) & 
       len(lat_list) == len(lng_list))

In [473]:
d = {
    'yelp_id' : id_list,
    'yelp_name' : name_list,
    'address' : address_list, 
    'is_closed' : is_closed_list,
    'categories' : categories_list,
    'price_level' : price_list,
    'avg_rating' : rating_list,
    'review_count' : review_count_list,
    'photo_urls' : photos_list,
    'latitude' : lat_list, 
    'longitude' : lng_list
}

df_yelp = pd.DataFrame(d)

In [474]:
#Filter out all restaurants that are  permanently closed -> before 920, afterwards 915
df_yelp.drop(df_yelp[df_yelp['is_closed']==True].index, inplace=True)

In [476]:
# Write details DataFrame to file
#df_yelp.to_csv('yelp_restaurants.csv', index=False)
#df_yelp.to_excel('yelp_restaurants.xlsx', index=False)

In [516]:
# For each Yelp ID, find the corresponding Google ID
# Do so by first looking up the Yelp name, then make a Google Places find_place search based on Yelp name and location

google_ids = []
google_names = []
yelp_name_list = df_yelp['yelp_name'].values

for yelp_name in yelp_name_list:
    google_restaurant = gmaps.find_place(
        input=yelp_name, input_type='textquery', 
        location_bias='point:' + str(lat_hbf) + ', ' + str(lng_hbf), 
        fields=['place_id', 'name'],
        language='de')
    print(yelp_name)
    if (google_restaurant['candidates']):
        google_id = google_restaurant['candidates'][0]['place_id']
        google_name = google_restaurant['candidates'][0]['name']
        print('...OK')
    else:
        google_id = ''
        google_name = ''
        print('...not found')
    google_ids.append(google_id)
    google_names.append(google_name)
        
assert(len(yelp_name_list) == len(google_ids) & len(google_ids) == len(google_names))

Bäckerhaus Veit Café
...OK
Prince of India
...OK
Bocca Buona
...OK
Weinwirtschaft | Weingut Franz Keller
...OK
Zum Adler Gasthaus Weinstube
...OK
ESV Rot-Weiss Vereinsgaststätte
...OK
Restaurant Schmidener Eintracht
...OK
Bäckerhaus Veit Café
...OK
Mikail's Lunchbox
...OK
India House
...OK
Da Mamma Maria
...OK
Ruppmannschänke
...OK
Schnauferl Grill
...OK
Goldener Adler
...OK
List
...OK
Taverne Diogenes
...OK
Wienerwald
...OK
Joe Peña's
...OK
Tschechen & Söhne
...OK
Dam Dam Hot Pizza
...OK
SEN Asia Soul Kitchen
...OK
Cannstatter Tor
...OK
Gaststätte Post
...OK
Kebap Stern
...OK
Geiler
...OK
Art Sushi
...OK
Wirtshaus zum Alten Bahnhof
...OK
Don Giovanni Ristorante Pizzeria
...OK
Cafe & Bar Nion
...OK
ALATURKA - Das Stuttgarter Original
...OK
Kwan Kao
...OK
Gasthof Franziska
...OK
Akeno
...OK
Noir Cuisine&Bar
...OK
Ristorante La Piazza
...OK
Christo's Stube
...OK
Bahtti Super Pizza Service
...OK
Cavallino Spaghettaro
...OK
Flori & Palma
...OK
Vereinsgaststätte Himmerreich
...OK
Restaurant

Nannina
...OK
Das Reiskorn
...OK
Feuerbacher Wichtel
...OK
Indian Shop
...OK
Hallo Emil
...OK
Brauereigaststätte Dinkelacker
...OK
Weinstube Am Stadtgraben
...OK
Ruben's
...OK
Galerie
...OK
Buschpilot
...OK
Die Palette
...OK
Nachtschicht
...OK
McDonald's
...OK
Christophorus
...OK
SGU Gaststätte Delphi
...OK
Cafe Wahnsinn
...OK
El Seco
...OK
Palazzo
...OK
QQ-Sushi Lounge
...OK
Distelfarm
...OK
Sultan Saray
...OK
Hegel Eins
...OK
Weinstube Viertelesschlotzer
...OK
Pizza & Pasta - San Marco
...OK
Ristorante La Scala
...OK
IL Cappuccino
...OK
POP's
...OK
Chicken Tom
...OK
Classic Rock Cafe
...OK
Kursaal Cannstatt
...OK
Weinstube Träuble
...OK
Aral
...OK
Restaurant Filoxenia
...OK
Chito Gvrito
...OK
Der Foehrich
...OK
Vier Burgen
...OK
Aslan Kebab
...OK
Restaurant Ochsen
...OK
Hirsch Gehrung
...OK
Merci
...OK
Udo-Snack
...OK
Zum Spätzleschwob
...OK
Sen Viet
...OK
Red Grill
...OK
Der Zauberlehrling
...OK
pizzabar
...OK
Piri Reis Restaurant & Cafe
...OK
Taverne Saloniki
...OK
KFC
...OK
Imbiss

Cannstatter Kebap Haus
...OK
Flora & Fauna
...OK
Cafebar Auszeit
...OK
Vereinsheim Gartenfreunde Heslach
...OK
Taverna Yol
...OK
Hans Im Glück
...OK
Restaurant Tsv Steinhaldenfeld
...OK
87 Barbecue & Bar
...OK
Reiterstüble Gasthaus
...OK
Restaurant Steinhalde
...OK
Wienerwald
...OK
Tagesbar
...OK
The Burger Republic
...OK
Restaurant Hermes
...OK
Seyffer's
...OK
Aral
...OK
Ristorante La Pergola
...OK
Burreatos
...OK
Gaststätte Ritterstüble
...OK
Trattoria Vivaldi
...OK
NLV Pizzeria
...OK
Jose y Joséfina
...OK
Cafe anno 1897
...OK
la bruschetta
...OK
Dedemoğlu
...OK
peperoncino
...OK
Sportrestaurant Neuwirtshaus
...OK
Restaurant Landhausstuben
...OK
Da Domenico
...OK
Hasen
...OK
Mögle
...OK
Injeera
...OK
DLRG Gaststätte
...OK
Max-Eyth-See
...OK
Gasthaus Rössle
...OK
Plieninger Most- und Weinbesen
...OK
Sapori d'Italia
...OK
Ilysia am Wallgraben
...OK
Persian Restaurant
...OK
Charisma Bootshaus
...OK
Osteria
...OK
Asia Imbiss Hop Sing
...OK
Rosenau - Lokalität & Bühne
...OK
Er Vaquita
...

In [518]:
df_yelp['google_id'] = google_ids
df_yelp['google_name'] = google_names

In [528]:
# Copy and re-order columns. 
df_yelp_reordered = df_yelp[['yelp_id', 'google_id', 'yelp_name', 'google_name', 'address', 'categories', 
                  'price_level', 'avg_rating', 'review_count', 'photo_urls', 'latitude', 'longitude']]
df_yelp = df_yelp_reordered

In [529]:
df_yelp.head()

yelp_id                    google_id  \
1  6XKiHruNrTO8jDAcHLoR3w  ChIJF5m2rSLFmUcRH9sBPHwx3kM   
2  0Ep-rCF4UVrz3ZJkd_gAXQ  ChIJOWEUrcvEmUcRqRXjZal8AZY   
3  MVScl0YGy_ZkGb7iN58k-A  ChIJI_1zBlHbmUcROMi0FydapcE   
4  KbIoL9bMYzCZDWxpWTcx0g  ChIJ6wNLZzPbmUcRrtanMZs8deo   
5  LlLDMhhPeCKs4oyZxFQGaw  ChIJo1Y9RTjQmUcRnsk3LcgHomE   

                               yelp_name  \
1                   Bäckerhaus Veit Café   
2                        Prince of India   
3                            Bocca Buona   
4  Weinwirtschaft | Weingut Franz Keller   
5           Zum Adler Gasthaus Weinstube   

                             google_name  \
1                   Bäckerhaus Veit Café   
2                        Prince of India   
3                 Bocca Buona Restaurant   
4  Weinwirtschaft | Weingut Franz Keller   
5               Hotel & Restaurant Adler   

                                             address  \
1          [Epplestr. 1 b, 70597 Stuttgart, Germany]   
2           [Urbanstr. 38, 70182 Stuttgart, Germany]   
3  [Hauptstätter Str. 147, 70180 Stuttgart, Germany]   
4        [Schillerstr. 23, 70173 Stuttgart, Germany]   
5  [Alte Stuttgarter Str. 2, 70195 Stuttgart, Ger...   

                           categories price_level  avg_rating  review_count  \
1                   [Bakeries, Cafes]                     1.5             5   
2                            [Indian]          €€         3.0            27   
3                           [Italian]                     2.5             3   
4  [International, German, Wine Bars]          €€         3.5            10   
5                       [Restaurants]           €         5.0             3   

                                          photo_urls   latitude  longitude  
1  [https://s3-media2.fl.yelpcdn.com/bphoto/M0vpn...  48.749109   9.169810  
2  [https://s3-media4.fl.yelpcdn.com/bphoto/rwxyn...  48.778664   9.187046  
3  [https://s3-media4.fl.yelpcdn.com/bphoto/NA4u_...  48.764643   9.169958  
4  [https://s3-media2.fl.yelpcdn.com/bphoto/l0icD...  48.782600   9.182460  
5  [https://s3-media2.fl.yelpcdn.com/bphoto/6_MBr...  48.778200   9.124740

In [ ]:
# Drop column 'google_name', as it may not be saved permanently
df_yelp = df_yelp.drop(columns=['google_name'])

In [530]:
# Write to file
df_yelp.to_csv('yelp_restaurants.csv', index=False)
df_yelp.to_excel('yelp_restaurants.xlsx', index=False)